File: desktop/entangler/src/main/java/ru/dbudyak/entangler/PropertiesWorker.java
package ru.dbudyak.entangler;

import javafx.scene.control.Button;
import javafx.scene.control.ListView;
import javafx.scene.control.SelectionMode;
import javafx.scene.control.cell.TextFieldListCell;
import javafx.scene.layout.AnchorPane;
import ru.dbudyak.entangler.models.base.BaseElement;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by dbudyak on 27.05.2014.
 */
@SuppressWarnings("ALL")
public class PropertiesWorker {

    private OnPropertiesUpdate propertiesUpdate;
    private AnchorPane props;
    private ListView keys, values;

    private HashMap<String, String> data = new HashMap<>();

    public HashMap<String, String> getElementData() {
        return data;
    }


    public void setOnPropertiesListener(OnPropertiesUpdate onPropertiesListener) {
        this.propertiesUpdate = onPropertiesListener;
    }

    public void setPropertiesLayout(AnchorPane lookup) {
        this.props = lookup;
        Button bDelete = (Button) this.props.lookup("#propBtnDelete");
        Button bRotate = (Button) this.props.lookup("#propBtnRotate");
        Button bFlip = (Button) this.props.lookup("#propBtnFlip");
        bDelete.setOnAction(event -> propertiesUpdate.onDelete());
        bRotate.setOnAction(event -> propertiesUpdate.onRotate());
        bFlip.setOnAction(event -> propertiesUpdate.onFlip());
        keys = (ListView) this.props.lookup("#propKeysList");
        values = (ListView) this.props.lookup("#propValuesList");
        values.setEditable(true);
        values.setCellFactory(TextFieldListCell.forListView());
        values.getSelectionModel().setSelectionMode(SelectionMode.SINGLE);
    }

    public AnchorPane getPropertiesAnchor() {
        return this.props;
    }

    public void setResult(String s) {
        data.put("Result", s);
    }

    public void setTop(String name, String isConnected) {
        data.put(name, String.valueOf(isConnected));
    }

    public void setRight(String name, String isConnected) {
        data.put(name, String.valueOf(isConnected));
    }

    public void setBottom(String name, String isConnected) {
        data.put(name, String.valueOf(isConnected));
    }

    public void setLeft(String name, String isConnected) {
        data.put(name, String.valueOf(isConnected));
    }

    public void setDetectorCounts() {
        data.put("Clicks", "measured");
    }

    public void setPropertiesVisibility(boolean visibility) {
        this.props.setVisible(visibility);
    }

    public void setName(String name) {
        data.put("Name", name);
    }

    public void setSourceOutputParam(double value) {
        data.put("Mode", "cos(" + String.valueOf(value) + ")|0> + sin(" + String.valueOf(value) + ")|1>");
    }

    public void setBSTheta(double val) {
        data.put("\u03B8", String.valueOf(val));
    }

    public void setType(BaseElement.ElementType type) {
    }

    public void updateProperties() {
        for (Map.Entry<String, String> e : data.entrySet()) {
            String key = e.getKey();
            String value = e.getValue();
            if (keys.getItems().contains(key)) {
                values.getItems().set(keys.getItems().indexOf(key), value);
            } else {
                keys.getItems().add(key);
                values.getItems().add(value);
            }
        }
    }

    public interface OnPropertiesUpdate {
        public void onDelete();

        public void onFlip();

        public void onRotate();
    }


}

File: desktop/entangler/src/main/kotlin/ru/dbudyak/entangler/PropertiesWorker.kt
package ru.dbudyak.entangler

import javafx.scene.control.Button
import javafx.scene.control.ListView
import javafx.scene.control.SelectionMode
import javafx.scene.control.cell.TextFieldListCell
import javafx.scene.layout.AnchorPane
import ru.dbudyak.entangler.models.BaseElement

/**
 * Manages the properties panel UI for quantum elements.
 * Singleton pattern for global access.
 */
object PropertiesWorker {
    private var propertiesUpdate: OnPropertiesUpdate? = null
    private var props: AnchorPane? = null
    private var keys: ListView<String>? = null
    private var values: ListView<String>? = null

    private val data = HashMap<String, String>()

    fun getElementData(): HashMap<String, String> = data

    fun setOnPropertiesListener(listener: OnPropertiesUpdate) {
        propertiesUpdate = listener
    }

    fun setPropertiesLayout(lookup: AnchorPane) {
        props = lookup
        val bDelete = props?.lookup("#propBtnDelete") as? Button
        val bRotate = props?.lookup("#propBtnRotate") as? Button
        val bFlip = props?.lookup("#propBtnFlip") as? Button

        bDelete?.setOnAction { propertiesUpdate?.onDelete() }
        bRotate?.setOnAction { propertiesUpdate?.onRotate() }
        bFlip?.setOnAction { propertiesUpdate?.onFlip() }

        keys = props?.lookup("#propKeysList") as? ListView<String>
        values = (props?.lookup("#propValuesList") as? ListView<String>)?.apply {
            isEditable = true
            cellFactory = TextFieldListCell.forListView()
            selectionModel.selectionMode = SelectionMode.SINGLE
        }
    }

    fun getPropertiesAnchor(): AnchorPane? = props

    fun setResult(s: String) {
        data["Result"] = s
    }

    fun setTop(name: String, isConnected: String) {
        data[name] = isConnected
    }

    fun setRight(name: String, isConnected: String) {
        data[name] = isConnected
    }

    fun setBottom(name: String, isConnected: String) {
        data[name] = isConnected
    }

    fun setLeft(name: String, isConnected: String) {
        data[name] = isConnected
    }

    fun setDetectorCounts() {
        data["Clicks"] = "measured"
    }

    fun setPropertiesVisibility(visibility: Boolean) {
        props?.isVisible = visibility
    }

    fun setName(name: String) {
        data["Name"] = name
    }

    fun setSourceOutputParam(value: Double) {
        data["Mode"] = "cos($value)|0> + sin($value)|1>"
    }

    fun setBSTheta(value: Double) {
        data["\u03B8"] = value.toString()
    }

    fun setType(type: BaseElement.ElementType) {
        // No-op for now
    }

    fun clearProperties() {
        keys?.items?.clear()
        values?.items?.clear()
    }

    fun clearData() {
        data.clear()
    }

    fun updateProperties() {
        clearProperties()
        data.forEach { (key, value) ->
            keys?.items?.add(key)
            values?.items?.add(value)
        }
    }

    /**
     * Interface for property update callbacks.
     */
    interface OnPropertiesUpdate {
        fun onDelete()
        fun onFlip()
        fun onRotate()
    }
}
